ipykernel/kernelbase.py

```diff
@@ -837,11 +835,12 @@ async def shutdown_request(self, stream, ident, parent):
 
         self.log.debug('Stopping control ioloop')
         control_io_loop = self.control_stream.io_loop
-        control_io_loop.add_callback(control_io_loop.stop)
+        control_io_loop.call_soon_threadsafe(control_io_loop.stop)
 
         self.log.debug('Stopping shell ioloop')
         shell_io_loop = self.shell_stream.io_loop
-        shell_io_loop.add_callback(shell_io_loop.stop)
+        # FIXME: shouldn't need to be threadsafe
+        shell_io_loop.call_soon_threadsafe(shell_io_loop.stop)
```

Comment:
I thought that we shouldn't need to `call_soon_threadsafe` because the shell event loop runs in the main thread, but actually we do because we are processing the shutdown request from the control channel, which runs in another thread.
Correct me if I'm wrong :smile:

Reply:
I think you're right